packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol

```diff
@@ -0,0 +1,136 @@
+     * liquidate a position within one transaction.
+     * @dev After the liquidation is done the DSProxy that called this method will have an open positon AND pending
+     * liquidation within the financial contract. The bot using the DSProxy should withdraw the liquidation once it has
+     * passed liveness. At this point the position can be manually unwound.
+     * @param uniswapRouter address of the uniswap router used to facilate trades.
+     * @param financialContract address of the financial contract on which the liquidation is occuring.
+     * @param reserveCurrency address of the token to swap for collateral. THis is the common currency held by the DSProxy.
+     * @param liquidatedSponsor address of the sponsor to be liquidated.
+     * @param maxReserverTokenSpent maximum number of reserve tokens to spend in the trade. Bounds slippage.
+     * @param minCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token is below this value.
+     * @param maxCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token exceeds this value.
+     * @param maxTokensToLiquidate max number of tokens to liquidate. For a full liquidation this is the full position debt.
+     * @param deadline abort the trade and liquidation if the transaction is mined after this timestamp.
+     **/
+    function swapMintLiquidate(
+        address uniswapRouter,
+        address financialContract,
+        address reserveCurrency,
+        address liquidatedSponsor,
+        FixedPoint.Unsigned calldata maxReserverTokenSpent,
+        FixedPoint.Unsigned calldata minCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxTokensToLiquidate,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+        FinancialContractInterface fc = FinancialContractInterface(financialContract);
+
+        // 1. Calculate how much collateral needed to mint maxTokensToLiquidate.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
+        FixedPoint.Unsigned memory collateralToMintAtGcr = maxTokensToLiquidate.mul(gcr);
+
+        // 2. Calculate how much collateral is needed for the final fee.
+        FinderInterface finder = FinderInterface(fc.finder());
+        StoreInterface store = StoreInterface(finder.getImplementationAddress("Store"));
+        FixedPoint.Unsigned memory totalCollateralNeeded =
+            collateralToMintAtGcr.add(store.computeFinalFee(fc.collateralCurrency()));
+
+        // 3. Swap reserve currency to get required collateral for the mint + final fee.
+        address[] memory path = new address[](2);
+        path[0] = reserveCurrency;
+        path[1] = fc.collateralCurrency();
+
+        TransferHelper.safeApprove(reserveCurrency, address(router), maxReserverTokenSpent.rawValue);
```

Comment:
nit: tiny tiny gas optimization we might consider down the road: we might want to just set this to max approval so we don't have to re-approve each time.

Note: I think this would only be better if we're not sure how many tokens the contract will pull. If the contract will pull _exactly_ the approved tokens, it is probably better to set it on each run since we would benefit from storage write netting: 0 -> nonzero -> 0 === 2 sloads (not two sstores).

Reply:
ye, that's a good point. the only downside is right now I actually sue this approval to limit the size of the trade. `maxReserverTokenSpent` bounds how many tokens can be spent!